File: retro/lowe/source/python/func/distance_function.py
import math
import ROOT
import macro

def distance_position_pmt(hitinfo,invalue):
    pmt = hitinfo.Get3Position()
    position = macro.Vector3(invalue[0],invalue[1],invalue[2])
    p_p = pmt - position
    return math.sqrt(p_p*p_p)

def distance_position_retro(hitinfo,invalue):
    positionx = invalue[0]
    positiony = invalue[1]
    positionz = invalue[2]
    pmtx = hitinfo.Get3Position().x
    pmty = hitinfo.Get3Position().y
    pmtz = hitinfo.Get3Position().z
    WCradius = hitinfo.WCradius
    sqrtpmt= math.sqrt(pmtx*pmtx + pmty*pmty)
    if WCradius < sqrtpmt:
        pmtx = pmtx*WCradius/sqrtpmt
        pmty = pmty*WCradius/sqrtpmt
    WClength = hitinfo.WClength
    half_WClength = WClength/2.
    a = (pmtx - positionx)*(pmtx - positionx) + (pmty - positiony) * (pmty - positiony)
    b = positionx * (positionx - pmtx) + positiony* (positiony - pmty)
    c = positionx * positionx + positiony * positiony - WCradius* WCradius
    t = 0.
    if a != 0.:
        t = (b-math.sqrt(b*b-a*c))/a
    else:
        t = -c/(2*b)
    z = positionz + (pmtz - positionz)*t
    if (-half_WClength < z ) and (half_WClength > z):
        pass
    elif z > half_WClength:
        t = (half_WClength - positionz)/(pmtz - positionz)
        z = half_WClength
    else:
        t = (-half_WClength - positionz)/(pmtz - positionz)
        z = - half_WClength
    x = positionx + (pmtx - positionx)*t
    y = positiony + (pmty - positiony)*t
    d2 = math.sqrt((positionx - x)*(positionx - x) + (positiony - y)*(positiony - y) + (positionz - z) * (positionz - z))
    return d2

def distance_fly_retro(hitinfo,invalue):
    d_p_p = distance_position_pmt(hitinfo,invalue)
    d_p_r = distance_position_retro(hitinfo,invalue)
    return (d_p_p + 2* d_p_r)

dfunctiondict = {}

File: retro/lowe/source/macro/funcparts/src/distance_function.cc
// c++ STL
#include <cmath>

// geant4 library
#include <CLHEP/Vector/ThreeVector.h>
// self-introduced library
#include "distance_function.hh"
double distance_position_pmt::returnvalue()
{
  CLHEP::Hep3Vector position_pmt = data.Get3Position() - info.Get3Position();
  double d = std::sqrt(position_pmt*position_pmt);
  return d;
}

double distance_position_retro::returnvalue()
{
  double positionX = data.Get3Position().x();
  double positionY = data.Get3Position().y();
  double positionZ = data.Get3Position().z();
  double pmtX = info.GetPosition(0);
  double pmtY = info.GetPosition(1);
  double pmtZ = info.GetPosition(2);
  double WCradius = info.GetWCradius();
  double WClength = info.GetWClength();
  double half_WClength = WClength/2.;
  double a = (pmtX-positionX)*(pmtX-positionX)+(pmtY-positionY)*(pmtY-positionY);
  double b = positionX*(positionX-pmtX)+positionY*(positionY-pmtY);
  double c = positionX*positionX+positionY*positionY-WCradius*WCradius;
  double t;
  if(a != 0){
    t = (b-sqrt(b*b-a*c))/a;
  }
  else{
    t = -c/(2*b);
  }
  double z = positionZ + (pmtZ - positionZ)*t;
  if(-half_WClength < z && half_WClength > z){

}
  else if(z > half_WClength){
    t = (half_WClength - positionX)/(pmtZ-positionX);
    z = half_WClength;
  }
  else{
    t = (-half_WClength - positionX)/(pmtZ-positionX);
    z = - half_WClength;
  }
  double x = positionX + (pmtX-positionX)*t;
  double y = positionY + (pmtY-positionY)*t;
  double d2 = std::sqrt((positionX-x)*(positionX-x)+(positionY-y)*(positionY-y)+(positionZ-z)*(positionZ-z));
  return d2;
}

double distance_fly_retro::returnvalue()
{
  distance_position_pmt d_position_pmt;
  d_position_pmt.SetHitInfo(info);
  d_position_pmt.SetReconstructdata(data);
  distance_position_retro d_position_retro;
  d_position_retro.SetHitInfo(info);
  d_position_retro.SetReconstructdata(data);
  double d_p_p = d_position_pmt.returnvalue();
  double d_p_r = d_position_retro.returnvalue();
  return (d_p_p + 2* d_p_r);
}

